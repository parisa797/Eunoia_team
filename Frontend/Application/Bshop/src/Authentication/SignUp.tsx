import React, { useRef, useState } from "react";
// import { TextInput as RNTextInput } from "react-native";
import { CommonActions } from "@react-navigation/native";
import { BorderlessButton } from "react-native-gesture-handler";
import { useFormik } from "formik";
import * as Yup from "yup";
import { Container, Button, Box } from "../components";
import { AuthNavigationProps } from "../components/Navigation";
// import TextInput from "../components/Form/TextInput";
import Checkbox from "../components/Form/Checkbox";
import Footer from "./components/Footer";

import { StatusBar } from "expo-status-bar";
import * as SecureStore from "expo-secure-store";
import FlashMessage from "react-native-flash-message";
import { showMessage, hideMessage } from "react-native-flash-message";
import {
  StyleSheet,
  Text,
  View,
  Image,
  TextInput,
  TouchableOpacity,
  Alert,
  TouchableWithoutFeedback,
  Keyboard,
} from "react-native";

const SignUp = ({ navigation }: AuthNavigationProps<"SignUp">) => {
  const [user, setUsername] = useState("");
  const [pass, setPassword] = useState("");
  const [confpass, setconferm_Password] = useState("");
  const [eml, setemail] = useState("");

  const password = useRef<RNTextInput>(null);
  const confirmPassword = useRef<RNTextInput>(null);

  const pressLogin = () => {
    console.log("im in kuft method");
    navigation.navigate("Login");
  };

  const validateEmail = (text) => {
    console.log("text is", text);
    let pattern = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
    return pattern.test(String(text).toLowerCase());
  };

  const validateUser = (text) => {
    console.log("text is", text);
    let pattern = /[a-zA-Z]/;
    return pattern.test(String(text).toLowerCase());
  };

  const HandleSignup = () => {
    if (
      user.lenght == 0 ||
      eml.length == 0 ||
      pass.length == 0 ||
      confpass.lengh == 0
    ) {
      showMessage({
        message: "لطفا اطلاعات کاربری خود را کامل وارد کنید.",
        type: "warning",
        backgroundColor: "#f1f1f2",
        color: "#000",
        statusBarHeight: "8",
        titleStyle: {
          fontSize: 15,
        },
      });
    } else {
      const requestOptions = {
        method: "POST",
        headers: {
          Accept: "application/json",
          "Content-Type": "application/json",
          "Access-Control-Allow-Origin": "*",
          "Access-Control-Allow-Headers": "*",
        },
        body: JSON.stringify({
          username: user,
          password1: pass,
          password2: confpass,
          email: eml,
        }),
      };
      fetch(
        "http://iust-bshop.herokuapp.com/rest-auth/registration/",
        requestOptions
      )
        .then((response) => {
          console.log("this is response", response);
          return response.json();
        })
        .then((result) => {
          console.log("this is result", result);
          var keys = Object.keys(result);
          if (keys.some((x) => x == "password1")) {
            alert(
              "توجه" +
                "\n" +
                "*رمز عبور باید شامل حداقل هشت کارکتر باشد." +
                "*رمز عبور شما باید شامل اعداد و حروف باشید." +
                "*رمز عبور باید پیچیده باشد.(لطفا از رمز عبور های رایج استفاده نکنید.)"
            );
          } else {
            if (keys.some((x) => x == "detail")) {
              alert(
                "ثبت نام شما با موفقیت انجام شد ! لطفا جهت ادامه و ورود ایمیل ارسالی را تایید کنید"
              );
              pressLogin();
            }
          }
        })
        .catch((error) => {
          console.log("something bad happened.", error);
        });
    }
  };
  return (
    <TouchableWithoutFeedback
      onPress={() => {
        Keyboard.dismiss();
      }}
    >
      <View style={styles.container}>
        {/* <Image style={styles.image} source={require("./assets/signup.png")}/> */}

        <View style={styles.inputView}>
          <TextInput
            value={user}
            style={styles.input}
            placeholder="نام کاربری"
            autoCapitalize="none"
            placeholderTextColor="#000"
            testID={"user_check"}
            onChangeText={(user) => setUsername(user)}
            onBlur={() => {
              const v = validateUser(user);
              console.log(v);
              if (v == false)
                alert(
                  "نام کاربری نمیتواند شامل ارقام باشد! لطفا از ترکیبی از حرف و عدد استفاده کنید."
                );
            }}
          />
        </View>
        <View style={styles.inputView}>
          <TextInput
            value={pass}
            style={styles.input}
            placeholder="رمز عبور"
            secureTextEntry={true}
            autoCapitalize="none"
            testID={"pass_check"}
            placeholderTextColor="#000"
            onChangeText={(pass) => setPassword(pass)}
          />
        </View>
        <View style={styles.inputView}>
          <TextInput
            value={confpass}
            style={styles.input}
            testID={"confpass_check"}
            placeholder="تکرار رمز عبور"
            secureTextEntry={true}
            autoCapitalize="none"
            placeholderTextColor="#000"
            onChangeText={(confpass) => setconferm_Password(confpass)}
            onBlur={() => {
              if (pass != confpass)
                alert("رمزعبورهای وارد شده باید با هم مطابقت داشته باشند.");
            }}
          />
        </View>
        <View style={styles.inputView}>
          <TextInput
            value={eml}
            style={styles.input}
            placeholder="ایمیل"
            autoCapitalize="none"
            testID={"eml_check"}
            placeholderTextColor="#000"
            onChangeText={(eml) => setemail(eml)}
            onBlur={() => {
              const v = validateEmail(eml);
              console.log(v);
              if (v == false) alert("لطفا ایمیل خود را درست وارد کنید.");
            }}
          />
        </View>
        <TouchableOpacity
          style={styles.SignUpBtn}
          onPress={() => HandleSignup()}
        >
          <Text style={styles.SignUpText}>ثبت نام</Text>
        </TouchableOpacity>
        <Text style={{ fontSize: 20 }}>حساب کاربری دارید؟</Text>
        <Text style={{ color: "#b31414", fontSize: 20 }} onPress={pressLogin}>
          ورود
        </Text>
      </View>
    </TouchableWithoutFeedback>
    // {/* </Container> */}
  );
};

const styles = StyleSheet.create({
  input: {
    width: 300,
    height: 55,
    backgroundColor: "#f1f1f2",
    marginBottom: 20,
    padding: 8,
    color: "black",
    borderRadius: 14,
    fontSize: 18,
    fontWeight: "500",
    textAlign: "center",
  },
  // downpage: {
  //   marginBottom: 40,
  // },
  container: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
    backgroundColor: "#fff",
    marginTop: "-5%",
  },

  TextInput: {
    height: 50,
    flex: 1,
    padding: 10,
    marginLeft: 20,
    fontSize: 20,
  },
  // image:{
  // width: "90%",
  // height:"50%",
  // },

  SignUpBtn: {
    width: 90,
    borderRadius: 10,
    height: 50,
    alignItems: "center",
    justifyContent: "center",
    marginTop: 20,
    marginBottom: 20,
    backgroundColor: "#b31414",
  },
});

export default SignUp;
