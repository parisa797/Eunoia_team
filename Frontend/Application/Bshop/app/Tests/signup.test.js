import React from "react";
import renderer from "react-test-renderer";
import { act } from "react-test-renderer";
import SignUp from "../signup";
// import { getByTestId } from "@testing-library/dom";
import {
  render,
  fireEvent,
  cleanup,
  waitFor,
} from "react-native-testing-library";
// import ReactDOM from "react-dom";
// import { Alert } from "react-native";
// import "jest-dom/extend-expect";

global.fetch = jest.fn();

// mocking an API success response once
fetch.mockResponseIsSuccess = (body) => {
  fetch.mockImplementationForOnce(() =>
    Promise.resolve({ json: () => Promise.resolve(JSON.parse(body)) })
  );
};

// mocking an API failure response for once
fetch.mockResponseIsFailure = (error) => {
  fetch.mockImplementationForOnce(() => Promise.reject(error));
};

describe("SignUp component be tested", () => {
  // afterEach(cleanup);

  it("renders correctly", () => {
    const tree = renderer.create(<SignUp />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("check email input", async () => {
    const push = jest.fn();
    const { getByTestId, getByPlaceholderText } = await render(
      <SignUp navigation={{ push }} />
    );
    const eml = "hi@gmail.com";
    await fireEvent.changeText(getByTestId("eml_check"), eml);
    expect(getByTestId("eml_check").props.value).toBe(eml);
  });

  it("check user_check input", async () => {
    const push = jest.fn();
    const { getByTestId, getByPlaceholderText } = await render(
      <SignUp navigation={{ push }} />
    );
    const user = "hojat";
    await fireEvent.changeText(getByTestId("user_check"), user);
    expect(getByTestId("user_check").props.value).toBe(user);
  });

  it("check pass_check input", async () => {
    const push = jest.fn();
    const { getByTestId, getByPlaceholderText } = await render(
      <SignUp navigation={{ push }} />
    );
    const pass = "13799731p";
    await fireEvent.changeText(getByTestId("pass_check"), pass);
    expect(getByTestId("pass_check").props.value).toBe(pass);
  });

  it("check confpass_check input", async () => {
    const push = jest.fn();
    const { getByTestId, getByPlaceholderText } = await render(
      <SignUp navigation={{ push }} />
    );
    const confpass = "13799731p";
    await fireEvent.changeText(getByTestId("confpass_check"), confpass);
    expect(getByTestId("confpass_check").props.value).toBe(confpass);
  });

  //   it("check email validation", async () => {
  //     const push = jest.fn();
  //     const { getByTestId, getByPlaceholderText } = await render(
  //       <SignUp navigation={{ push }} />
  //     );

  //     errorMessageText = "لطفا ایمیل خود را درست وارد کنید.";
  //     Alert.alert = jest.fn().mockResolvedValue(errorMessageText);
  //     // jest.spyOn(Alert, "alert");

  //     const emailvalidation = "phj.com";
  //     await fireEvent.changeText(getByTestId("eml_check"), emailvalidation);
  //     // expect(getByTestId("eml_check").props.value).toBe(emailvalidation);
  //     // expect(Alert.alert).toHaveBeenCalledWith(errorMessageText);
  //   });

  //   it("check same pass and confpass", async () => {
  //     const push = jest.fn();
  //     const { getByTestId, getByPlaceholderText } = await render(
  //       <SignUp navigation={{ push }} />
  //     );

  //     errorMessageText = "رمزعبورهای وارد شده باید با هم مطابقت داشته باشند.";
  //     Alert.alert = jest.fn().mockResolvedValue(errorMessageText);
  //     // jest.spyOn(Alert, "alert");
  //     const pass = "13799731p";
  //     await fireEvent.changeText(getByTestId("pass_check"), pass);
  //     const samepass = "13799731pp";
  //     await fireEvent.changeText(getByTestId("confpass_check"), samepass);
  //     // expect(getByTestId("eml_check").props.value).toBe(emailvalidation);
  //     // expect(Alert.alert).toHaveBeenCalledWith(errorMessageText);
  //   });

  //   it("check username has number and alphabet", async () => {
  //     const push = jest.fn();
  //     const { getByTestId, getByPlaceholderText } = await render(
  //       <SignUp navigation={{ push }} />
  //     );

  //     errorMessageText =
  //       "نام کاربری نمیتواند شامل ارقام باشد! لطفا از ترکیبی از حرف و عدد استفاده کنید.";
  //     Alert.alert = jest.fn().mockResolvedValue(errorMessageText);
  //     // jest.spyOn(Alert, "alert");

  //     const cheakboathuserform = "1254";
  //     await fireEvent.changeText(getByTestId("user_check"), cheakboathuserform);
  //     // expect(getByTestId("eml_check").props.value).toBe(emailvalidation);
  //     // expect(Alert.alert).toHaveBeenCalledWith(errorMessageText);
  //   });
});
