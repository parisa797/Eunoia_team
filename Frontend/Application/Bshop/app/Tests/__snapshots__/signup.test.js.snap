// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp component be tested renders correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
      "marginTop": "-5%",
    }
  }
>
  <Image
    source={1}
    style={
      Object {
        "height": "50%",
        "marginTop": -35,
        "width": "80%",
      }
    }
  />
  <View>
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      onBlur={[Function]}
      onChangeText={[Function]}
      placeholder="نام کاربری"
      placeholderTextColor="#000"
      rejectResponderTermination={true}
      style={
        Object {
          "backgroundColor": "#f1f1f2",
          "borderRadius": 14,
          "color": "black",
          "fontSize": 18,
          "fontWeight": "500",
          "height": 55,
          "marginBottom": 20,
          "marginTop": -10,
          "padding": 8,
          "textAlign": "center",
          "width": 300,
        }
      }
      testID="user_check"
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <View>
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      onChangeText={[Function]}
      placeholder="رمز عبور"
      placeholderTextColor="#000"
      rejectResponderTermination={true}
      secureTextEntry={true}
      style={
        Object {
          "backgroundColor": "#f1f1f2",
          "borderRadius": 14,
          "color": "black",
          "fontSize": 18,
          "fontWeight": "500",
          "height": 55,
          "marginBottom": 20,
          "marginTop": -10,
          "padding": 8,
          "textAlign": "center",
          "width": 300,
        }
      }
      testID="pass_check"
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <View>
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      onBlur={[Function]}
      onChangeText={[Function]}
      placeholder="تکرار رمز عبور"
      placeholderTextColor="#000"
      rejectResponderTermination={true}
      secureTextEntry={true}
      style={
        Object {
          "backgroundColor": "#f1f1f2",
          "borderRadius": 14,
          "color": "black",
          "fontSize": 18,
          "fontWeight": "500",
          "height": 55,
          "marginBottom": 20,
          "marginTop": -10,
          "padding": 8,
          "textAlign": "center",
          "width": 300,
        }
      }
      testID="confpass_check"
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <View>
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      onBlur={[Function]}
      onChangeText={[Function]}
      placeholder="ایمیل"
      placeholderTextColor="#000"
      rejectResponderTermination={true}
      style={
        Object {
          "backgroundColor": "#f1f1f2",
          "borderRadius": 14,
          "color": "black",
          "fontSize": 18,
          "fontWeight": "500",
          "height": 55,
          "marginBottom": 20,
          "marginTop": -10,
          "padding": 8,
          "textAlign": "center",
          "width": 300,
        }
      }
      testID="eml_check"
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#b31414",
        "borderRadius": 10,
        "height": 50,
        "justifyContent": "center",
        "marginBottom": 20,
        "marginTop": 10,
        "opacity": 1,
        "width": 90,
      }
    }
  >
    <Text>
      ثبت نام
    </Text>
  </View>
  <Text
    onPress={[Function]}
    style={
      Object {
        "color": "#b31414",
        "fontSize": 20,
        "marginTop": -15,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "black",
          "fontSize": 20,
        }
      }
    >
      حساب کاربری دارید؟
    </Text>
    ورود
  </Text>
</View>
`;
